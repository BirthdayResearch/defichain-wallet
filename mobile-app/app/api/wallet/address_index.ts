import { SecuredStoreAPI } from '@api'

const KEY = 'WALLET_ADDRESS.INDEX'

async function getLength (): Promise<number> {
  const str = await SecuredStoreAPI.getItem(`${KEY}.length`)
  return str === undefined ? 0 : Number(str)
}

async function setLength (count: number): Promise<void> {
  await SecuredStoreAPI.setItem(`${KEY}.length`, `${count}`)
}

async function getActive (): Promise<number> {
  const str = await SecuredStoreAPI.getItem(`${KEY}.active`)
  return str === undefined ? 0 : Number(str)
}

async function setActive (count: number): Promise<void> {
  await SecuredStoreAPI.setItem(`${KEY}.active`, `${count}`)
}

/**
 * max index to know how many addresses have been generated.
 * active index to know which index's address is active.
 */
export const WalletAddressIndexPersistence = {
  getLength,
  setLength,
  getActive,
  setActive
}
