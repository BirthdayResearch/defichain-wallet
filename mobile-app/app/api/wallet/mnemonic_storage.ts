import { PrivateKeyEncryption, Scrypt } from '@defichain/jellyfish-wallet-encrypted'
import { entropyAsMnemonic, mnemonicAsEntropy } from '@defichain/jellyfish-wallet-mnemonic'
import { getRandomBytes } from 'expo-random'
import { SecuredStoreAPI } from '@api'

const KEY = 'ENCRYPTED_MNEMONIC_STORAGE.entropy'

/**
 * Raw mnemonic words encryption implementation reside in light wallet
 */
class EncryptedMnemonicStorage {
  /**
   * jellyfish's PrivateKeyEncryption impl essentially work for infinite length of data
   */
  private readonly encryption: PrivateKeyEncryption

  constructor () {
    this.encryption = new PrivateKeyEncryption(new Scrypt(), numOfBytes => {
      const bytes = getRandomBytes(numOfBytes)
      return Buffer.from(bytes)
    })
  }

  /**
   * Encrypt mnemonic words, and store into persistent storage.
   *
   * @param {string[]} words
   * @param {string} passphrase
   */
  async set (words: string[], passphrase: string): Promise<void> {
    const buffer = mnemonicAsEntropy(words)
    const encryptedData = await this.encryption.encrypt(buffer, passphrase)
    const encoded = encryptedData.encode()
    await SecuredStoreAPI.setItem(KEY, encoded)
  }

  /**
   * Retrieve encrypted value from secure store, deserialize, decrypt, deserialize buffer (decryption result) into words array
   *
   * @param {string} passphrase
   * @returns {string[]}
   */
  async get (passphrase: string): Promise<string[]> {
    const encrypted = await SecuredStoreAPI.getItem(KEY)
    if (encrypted === null || encrypted === undefined) {
      throw new Error('NO_MNEMONIC_BACKUP')
    }

    const buffer = await this.encryption.decrypt(encrypted, passphrase)
    return entropyAsMnemonic(buffer)
  }
}

export const MnemonicStorage = new EncryptedMnemonicStorage()
