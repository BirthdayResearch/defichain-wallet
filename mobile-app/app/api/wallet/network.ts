import { MainNet, Network, RegTest, TestNet } from '@defichain/jellyfish-network'
import { Bip32Options } from '@defichain/jellyfish-wallet-mnemonic'
import { EnvironmentNetwork } from '@environment'

export function getJellyfishNetwork (network: EnvironmentNetwork): Network {
  switch (network) {
    case EnvironmentNetwork.MainNet:
      return MainNet
    case EnvironmentNetwork.TestNet:
      return TestNet
    case EnvironmentNetwork.LocalPlayground:
    case EnvironmentNetwork.RemotePlayground:
      return RegTest
  }
}

export function getBip32Option (envNetwork: EnvironmentNetwork): Bip32Options {
  const network = getJellyfishNetwork(envNetwork)
  return {
    bip32: {
      public: network.bip32.publicPrefix,
      private: network.bip32.privatePrefix
    },
    wif: network.wifPrefix
  }
}
