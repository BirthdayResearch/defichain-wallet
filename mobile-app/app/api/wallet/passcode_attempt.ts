import { SecuredStoreAPI } from '@api'

const KEY = 'PASSCODE_ATTEMPT.count'

async function get (): Promise<number> {
  const str = await SecuredStoreAPI.getItem(KEY)
  return str === undefined ? 0 : Number(str)
}

async function set (count: number): Promise<void> {
  await SecuredStoreAPI.setItem(KEY, `${count}`)
}

/**
 * Failed passcode input counter persistence layer
 */
export const PasscodeAttemptCounter = {
  set,
  get
}
