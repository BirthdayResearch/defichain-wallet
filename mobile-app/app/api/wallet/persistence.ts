import { SecuredStoreAPI } from '@api'
import { WalletPersistenceDataI } from '@shared-contexts/WalletPersistenceContext'

async function get (): Promise<Array<WalletPersistenceDataI<any>>> {
  const count: string = await SecuredStoreAPI.getItem('WALLET.count') ?? '0'

  const list: Array<WalletPersistenceDataI<any>> = []
  for (let i = 0; i < parseInt(count); i++) {
    const data = await SecuredStoreAPI.getItem(`WALLET.${i}`)
    if (data === null) {
      throw new Error(`WALLET.count=${count} but ${i} doesn't exist`)
    }

    list[i] = JSON.parse(data)
  }
  return list
}

/**
 * @param wallets to set, override previous set wallet
 */
async function set (wallets: Array<WalletPersistenceDataI<any>>): Promise<void> {
  await clear()

  for (let i = 0; i < wallets.length; i++) {
    await SecuredStoreAPI.setItem(`WALLET.${i}`, JSON.stringify(wallets[i]))
  }
  await SecuredStoreAPI.setItem('WALLET.count', `${wallets.length}`)
}

/**
 * Clear all persisted wallet
 */
async function clear (): Promise<void> {
  const count: string = await SecuredStoreAPI.getItem('WALLET.count') ?? '0'
  for (let i = 0; i < parseInt(count); i++) {
    await SecuredStoreAPI.removeItem(`WALLET.${i}`)
  }
}

/**
 * Multi Wallet Persistence Layer
 */
export const WalletPersistence = {
  set,
  get
}
