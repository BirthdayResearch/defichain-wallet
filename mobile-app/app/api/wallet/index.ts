import { JellyfishWallet, WalletHdNode, WalletHdNodeProvider } from '@defichain/jellyfish-wallet'
import { WhaleApiClient } from '@defichain/whale-api-client'
import { WhaleWalletAccount, WhaleWalletAccountProvider } from '@defichain/whale-api-wallet'
import { EnvironmentNetwork } from '@environment'
import { getJellyfishNetwork } from '@shared-api/wallet/network'

export function initJellyfishWallet (provider: WalletHdNodeProvider<WalletHdNode>, network: EnvironmentNetwork, client: WhaleApiClient): JellyfishWallet<WhaleWalletAccount, WalletHdNode> {
  const accountProvider = new WhaleWalletAccountProvider(client, getJellyfishNetwork(network))
  return new JellyfishWallet(provider, accountProvider)
}

export * from './provider/mnemonic_encrypted'
export * from './provider/mnemonic_unprotected'
export * from './passcode_attempt'
export * from './persistence'
