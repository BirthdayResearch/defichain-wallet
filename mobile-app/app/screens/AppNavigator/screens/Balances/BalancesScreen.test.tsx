import { configureStore } from '@reduxjs/toolkit'
import { fireEvent, render } from '@testing-library/react-native'
import { Provider } from 'react-redux'
import { RootState } from '@store'
import { wallet, setTokenSymbol } from '@store/wallet'
import { block } from '@store/block'
import { BalancesScreen } from './BalancesScreen'

jest.mock('@react-navigation/bottom-tabs', () => ({
  useBottomTabBarHeight: () => 49
}))
jest.mock('randomcolor', () => jest.fn().mockReturnValue('#ffffff'))
jest.mock('@shared-contexts/ThemeProvider')
jest.mock('@shared-contexts/LanguageProvider')
jest.mock('@shared-contexts/DeFiScanContext')
jest.mock('@shared-contexts/WalletContext')
jest.mock('@shared-contexts/WalletPersistenceContext')
jest.mock('@shared-contexts/NetworkContext')
jest.mock('@contexts/DisplayBalancesContext')

jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn()
}))
jest.mock('@gorhom/bottom-sheet', () => ({
  useBottomSheetModal: () => ({
    dismiss: jest.fn()
  })
}))

jest.mock('react-native/Libraries/Utilities/Platform', () => ({
  OS: 'web',
  select: () => jest.fn
}))

describe('balances page', () => {
  const tokens = [{
    id: '0',
    symbol: 'DFI',
    symbolKey: 'DFI',
    displaySymbol: 'DFI',
    isDAT: true,
    isLPS: false,
    isLoanToken: false,
    amount: '23',
    name: 'Defi'
  }, {
    id: '1',
    symbol: 'BTC',
    symbolKey: 'BTC',
    displaySymbol: 'dBTC',
    isDAT: true,
    isLPS: false,
    isLoanToken: false,
    amount: '777',
    name: 'Bitcoin'
  },
  {
    id: '2',
    symbol: 'ETH',
    symbolKey: 'ETH',
    displaySymbol: 'dETH',
    isDAT: true,
    isLPS: false,
    isLoanToken: false,
    amount: '555',
    name: 'Ethereum'
  }]

  it('should match snapshot', async () => {
    const initialState: Partial<RootState> = {
      wallet: {
        utxoBalance: '77',
        tokens: tokens.map(setTokenSymbol),
        allTokens: {},
        poolpairs: [],
        hasFetchedPoolpairData: false
      },
      block: {
        count: 100,
        masternodeCount: 10,
        lastSuccessfulSync: undefined,
        connected: true,
        isPolling: true,
        tvl: undefined
      }
    }
    const store = configureStore({
      preloadedState: initialState,
      reducer: {
        wallet: wallet.reducer,
        block: block.reducer
      }
    })
    const navigation: any = {
      navigate: jest.fn()
    }
    const route: any = {}
    const component = (
      <Provider store={store}>
        <BalancesScreen
          navigation={navigation}
          route={route}
        />
      </Provider>
    )
    const rendered = render(component)
    expect(rendered.toJSON()).toMatchSnapshot()
  })

  it('should navigate to token detail page', async () => {
    const initialState: Partial<RootState> = {
      wallet: {
        utxoBalance: '77',
        tokens: tokens.map(setTokenSymbol),
        allTokens: {},
        poolpairs: [],
        hasFetchedPoolpairData: false
      },
      block: {
        count: 100,
        masternodeCount: 10,
        lastSuccessfulSync: undefined,
        connected: true,
        isPolling: true,
        tvl: undefined
      }
    }
    const store = configureStore({
      preloadedState: initialState,
      reducer: {
        wallet: wallet.reducer,
        block: block.reducer
      }
    })
    const navigation: any = {
      navigate: jest.fn()
    }
    const route: any = {}
    const spy = jest.spyOn(navigation, 'navigate')
    const component = (
      <Provider store={store}>
        <BalancesScreen
          navigation={navigation}
          route={route}
        />
      </Provider>
    )
    const rendered = render(component)
    const btcBalanceRow = await rendered.findByTestId('balances_row_1')
    fireEvent.press(btcBalanceRow)
    expect(spy).toHaveBeenCalled()
  })
})
