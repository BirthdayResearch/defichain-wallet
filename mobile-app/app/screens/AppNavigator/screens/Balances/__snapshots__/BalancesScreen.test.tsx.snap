// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`balances page should match snapshot 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  ListHeaderComponent={
    <React.Fragment>
      <DFIBalanceCard />
      <ThemedSectionTitle
        testID="balances_title"
        text="PORTFOLIO"
      />
    </React.Fragment>
  }
  data={
    Array [
      Object {
        "amount": "777",
        "id": "1",
        "isDAT": true,
        "isLPS": false,
        "name": "Bitcoin",
        "symbol": "BTC",
        "symbolKey": "BTC",
      },
      Object {
        "amount": "555",
        "id": "2",
        "isDAT": true,
        "isLPS": false,
        "name": "Ethereum",
        "symbol": "ETH",
        "symbolKey": "ETH",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Array [
      undefined,
      Object {
        "backgroundColor": "rgba(245, 245, 245, 1)",
      },
    ]
  }
  testID="balances_list"
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Array [
            Object {
              "borderBottomLeftRadius": 8,
              "borderBottomRightRadius": 8,
              "borderTopLeftRadius": 8,
              "borderTopRightRadius": 8,
              "flexBasis": "0%",
              "flexGrow": 1,
              "flexShrink": 1,
              "marginLeft": 8,
              "marginRight": 8,
              "marginTop": 16,
            },
            Object {
              "backgroundColor": "rgba(255, 255, 255, 1)",
              "borderBottomColor": "rgba(245, 245, 245, 1)",
              "borderLeftColor": "rgba(245, 245, 245, 1)",
              "borderRightColor": "rgba(245, 245, 245, 1)",
              "borderTopColor": "rgba(245, 245, 245, 1)",
            },
          ]
        }
        testID="dfi_balance_card"
      >
        <View
          accessibilityIgnoresInvertColors={true}
          style={
            Object {
              "borderBottomLeftRadius": 8,
              "borderBottomRightRadius": 8,
              "borderTopLeftRadius": 8,
              "borderTopRightRadius": 8,
              "flexBasis": "0%",
              "flexGrow": 1,
              "flexShrink": 1,
              "overflow": "hidden",
            }
          }
        >
          <Image
            resizeMethod="scale"
            resizeMode="cover"
            source={1}
            style={
              Array [
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                Object {
                  "height": undefined,
                  "width": undefined,
                },
                undefined,
              ]
            }
          />
          <View
            style={
              Object {
                "flexBasis": "0%",
                "flexDirection": "column",
                "flexGrow": 1,
                "flexShrink": 1,
                "marginBottom": 16,
                "marginLeft": 16,
                "marginRight": 16,
                "marginTop": 20,
              }
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "marginBottom": 12,
                    }
                  }
                >
                  <RNSVGSvgView
                    align="xMidYMid"
                    bbHeight={24}
                    bbWidth={24}
                    focusable={false}
                    height={24}
                    meetOrSlice={0}
                    minX={0}
                    minY={0}
                    style={
                      Array [
                        Object {
                          "backgroundColor": "transparent",
                          "borderWidth": 0,
                        },
                        Object {
                          "marginRight": 8,
                        },
                        Object {
                          "flex": 0,
                          "height": 24,
                          "width": 24,
                        },
                      ]
                    }
                    vbHeight={32}
                    vbWidth={32}
                    width={24}
                  >
                    <RNSVGGroup>
                      <RNSVGCircle
                        cx={16}
                        cy={16}
                        fill={4294901935}
                        propList={
                          Array [
                            "fill",
                          ]
                        }
                        r={16}
                      />
                      <RNSVGPath
                        d="M18.217 23.454V8.546c3.21.958 5.555 3.937 5.555 7.454s-2.346 6.496-5.555 7.454M15.994 6v8.428l-1.269-1.27-.162-3.167 1.324-3.985a9.952 9.952 0 00-2.454.338l-.64 1.926-1.816-.91a10.03 10.03 0 00-1.974 1.497l3.363 1.685.098 1.928-1.927-.099L8.85 9.01a10.035 10.035 0 00-1.497 1.973l.911 1.816-1.927.64A9.978 9.978 0 006 15.892l3.986-1.324 3.168.162 1.27 1.27-1.27 1.27-3.168.162L6 16.108c.009.849.13 1.669.338 2.454l1.927.64-.911 1.816c.418.719.92 1.383 1.497 1.973l1.686-3.362 1.927-.099-.098 1.928-3.363 1.685a10.04 10.04 0 001.974 1.498l1.815-.911.64 1.926a9.94 9.94 0 002.455.338l-1.324-3.985.162-3.168 1.27-1.27V26c5.522 0 10-4.477 10-10s-4.478-10-10-10"
                        fill={4294967295}
                        propList={
                          Array [
                            "fill",
                          ]
                        }
                      />
                    </RNSVGGroup>
                  </RNSVGSvgView>
                  <Text
                    style={
                      Array [
                        Object {
                          "fontFamily": "RegularFont",
                          "fontSize": 16,
                          "fontWeight": "400",
                          "lineHeight": 24,
                        },
                        Array [
                          Object {
                            "fontFamily": "BoldFont",
                            "fontSize": 18,
                            "fontWeight": "700",
                            "lineHeight": 28,
                            "paddingRight": 36,
                          },
                          Object {
                            "color": "rgba(0, 0, 0, 1)",
                          },
                        ],
                      ]
                    }
                    testID="total_dfi_label"
                  >
                    DFI
                  </Text>
                </View>
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "RegularFont",
                        "fontSize": 16,
                        "fontWeight": "400",
                        "lineHeight": 24,
                      },
                      Array [
                        Object {
                          "fontSize": 14,
                          "lineHeight": 20,
                          "paddingBottom": 6,
                          "paddingRight": 56,
                        },
                        Object {
                          "color": "rgba(115, 115, 115, 1)",
                        },
                      ],
                    ]
                  }
                  testID="dfi_utxo_label"
                >
                  UTXO
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "RegularFont",
                        "fontSize": 16,
                        "fontWeight": "400",
                        "lineHeight": 24,
                      },
                      Array [
                        Object {
                          "fontSize": 14,
                          "lineHeight": 20,
                          "paddingRight": 48,
                        },
                        Object {
                          "color": "rgba(115, 115, 115, 1)",
                        },
                      ],
                    ]
                  }
                  testID="dfi_token_label"
                >
                  Token
                </Text>
              </View>
              <View
                style={
                  Object {
                    "paddingTop": 2,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "RegularFont",
                        "fontSize": 16,
                        "fontWeight": "400",
                        "lineHeight": 24,
                      },
                      Array [
                        Object {
                          "paddingBottom": 14,
                        },
                        Object {
                          "color": "rgba(0, 0, 0, 1)",
                        },
                      ],
                    ]
                  }
                  testID="total_dfi_amount"
                >
                  77.00000000
                   DFI
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "RegularFont",
                        "fontSize": 16,
                        "fontWeight": "400",
                        "lineHeight": 24,
                      },
                      Array [
                        Object {
                          "fontSize": 14,
                          "lineHeight": 20,
                          "paddingBottom": 6,
                        },
                        Object {
                          "color": "rgba(115, 115, 115, 1)",
                        },
                      ],
                    ]
                  }
                  testID="dfi_utxo_amount"
                >
                  77.00000000
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "RegularFont",
                        "fontSize": 16,
                        "fontWeight": "400",
                        "lineHeight": 24,
                      },
                      Array [
                        Object {
                          "fontSize": 14,
                          "lineHeight": 20,
                        },
                        Object {
                          "color": "rgba(115, 115, 115, 1)",
                        },
                      ],
                    ]
                  }
                  testID="dfi_token_amount"
                >
                  0.00000000
                </Text>
              </View>
            </View>
            <View
              style={
                Object {
                  "flexDirection": "row",
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "flex-end",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "opacity": 1,
                    "width": "75%",
                  }
                }
                testID="token_vs_utxo_info"
              >
                <Text />
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "RegularFont",
                        "fontSize": 16,
                        "fontWeight": "400",
                        "lineHeight": 24,
                      },
                      Array [
                        Object {
                          "fontFamily": "MediumFont",
                          "fontSize": 12,
                          "fontWeight": "500",
                          "lineHeight": 16,
                          "marginLeft": 4,
                          "paddingLeft": 4,
                          "paddingRight": 4,
                        },
                        Object {
                          "color": "rgba(255, 0, 175, 1)",
                        },
                      ],
                    ]
                  }
                >
                  Learn more about UTXO and tokens
                </Text>
              </View>
              <View
                style={
                  Object {
                    "flexDirection": "row",
                    "flexGrow": 1,
                    "justifyContent": "flex-end",
                  }
                }
              >
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "rgba(255, 255, 255, 1)",
                      "borderBottomColor": "rgba(212, 212, 212, 1)",
                      "borderBottomLeftRadius": 4,
                      "borderBottomRightRadius": 4,
                      "borderBottomWidth": 1,
                      "borderLeftColor": "rgba(212, 212, 212, 1)",
                      "borderLeftWidth": 1,
                      "borderRightColor": "rgba(212, 212, 212, 1)",
                      "borderRightWidth": 1,
                      "borderTopColor": "rgba(212, 212, 212, 1)",
                      "borderTopLeftRadius": 4,
                      "borderTopRightRadius": 4,
                      "borderTopWidth": 1,
                      "flexDirection": "row",
                      "marginRight": 8,
                      "opacity": 1,
                      "paddingBottom": 4,
                      "paddingLeft": 4,
                      "paddingRight": 4,
                      "paddingTop": 4,
                    }
                  }
                  testID="convert_dfi_button"
                >
                  <Text />
                </View>
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "rgba(255, 255, 255, 1)",
                      "borderBottomColor": "rgba(212, 212, 212, 1)",
                      "borderBottomLeftRadius": 4,
                      "borderBottomRightRadius": 4,
                      "borderBottomWidth": 1,
                      "borderLeftColor": "rgba(212, 212, 212, 1)",
                      "borderLeftWidth": 1,
                      "borderRightColor": "rgba(212, 212, 212, 1)",
                      "borderRightWidth": 1,
                      "borderTopColor": "rgba(212, 212, 212, 1)",
                      "borderTopLeftRadius": 4,
                      "borderTopRightRadius": 4,
                      "borderTopWidth": 1,
                      "flexDirection": "row",
                      "opacity": 1,
                      "paddingBottom": 4,
                      "paddingLeft": 4,
                      "paddingRight": 4,
                      "paddingTop": 4,
                    }
                  }
                  testID="send_dfi_button"
                >
                  <Text />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <Text
        style={
          Array [
            Object {
              "fontFamily": "RegularFont",
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
            Array [
              Object {
                "color": "rgba(115, 115, 115, 1)",
                "fontFamily": "MediumFont",
                "fontSize": 12,
                "fontWeight": "500",
                "lineHeight": 16,
                "paddingBottom": 8,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 24,
              },
              Object {
                "backgroundColor": "rgba(245, 245, 245, 1)",
                "color": "rgba(115, 115, 115, 1)",
              },
            ],
          ]
        }
        testID="balances_title"
        text="PORTFOLIO"
      >
        PORTFOLIO
      </Text>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(255, 255, 255, 1)",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "opacity": 1,
            "paddingBottom": 16,
            "paddingLeft": 16,
            "paddingRight": 8,
            "paddingTop": 16,
          }
        }
        testID="balances_row_1"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "flexGrow": 1,
            }
          }
        >
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight={32}
            bbWidth={32}
            focusable={false}
            height={32}
            meetOrSlice={0}
            minX={0}
            minY={0}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "flex": 0,
                  "height": 32,
                  "width": 32,
                },
              ]
            }
            testID="balances_row_1_icon"
            vbHeight={32}
            vbWidth={32}
            width={32}
          >
            <RNSVGGroup>
              <RNSVGCircle
                cx={16}
                cy={16}
                fill={4294967295}
                propList={
                  Array [
                    "fill",
                  ]
                }
                r={16}
              />
              <RNSVGText
                alignmentBaseline="central"
                content={null}
                dx={Array []}
                dy={Array []}
                fill={4294967295}
                font={
                  Object {
                    "fontSize": "24",
                    "fontWeight": "bolder",
                    "textAnchor": "middle",
                  }
                }
                propList={
                  Array [
                    "fill",
                  ]
                }
                rotate={Array []}
                x={
                  Array [
                    "50%",
                  ]
                }
                y={
                  Array [
                    "50%",
                  ]
                }
              >
                <RNSVGTSpan
                  content="T"
                  dx={Array []}
                  dy={Array []}
                  font={Object {}}
                  rotate={Array []}
                  x={Array []}
                  y={Array []}
                />
              </RNSVGText>
            </RNSVGGroup>
          </RNSVGSvgView>
          <View
            style={
              Object {
                "flexBasis": "auto",
                "flexGrow": 1,
                "flexShrink": 1,
                "marginLeft": 12,
                "marginRight": 12,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "RegularFont",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "lineHeight": 24,
                  },
                  Array [
                    Object {
                      "fontFamily": "MediumFont",
                      "fontWeight": "500",
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 1)",
                    },
                  ],
                ]
              }
              testID="balances_row_1_symbol"
            />
            <Text
              ellipsizeMode="tail"
              numberOfLines={1}
              style={
                Array [
                  Object {
                    "fontFamily": "RegularFont",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "lineHeight": 24,
                  },
                  Array [
                    Object {
                      "color": "rgba(82, 82, 82, 1)",
                      "fontFamily": "MediumFont",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "lineHeight": 20,
                    },
                    Object {
                      "color": "rgba(82, 82, 82, 1)",
                    },
                  ],
                ]
              }
              testID="balances_row_1_name"
            >
              Bitcoin
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "RegularFont",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "lineHeight": 24,
                  },
                  Array [
                    Object {
                      "flexWrap": "wrap",
                      "marginRight": 8,
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 1)",
                    },
                  ],
                ]
              }
              testID="balances_row_1_amount"
            >
              777.00000000
            </Text>
            <Text />
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "height": 1,
            },
            Object {
              "backgroundColor": "rgba(245, 245, 245, 1)",
            },
          ]
        }
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(255, 255, 255, 1)",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "opacity": 1,
            "paddingBottom": 16,
            "paddingLeft": 16,
            "paddingRight": 8,
            "paddingTop": 16,
          }
        }
        testID="balances_row_2"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "flexGrow": 1,
            }
          }
        >
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight={32}
            bbWidth={32}
            focusable={false}
            height={32}
            meetOrSlice={0}
            minX={0}
            minY={0}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "flex": 0,
                  "height": 32,
                  "width": 32,
                },
              ]
            }
            testID="balances_row_2_icon"
            vbHeight={32}
            vbWidth={32}
            width={32}
          >
            <RNSVGGroup>
              <RNSVGCircle
                cx={16}
                cy={16}
                fill={4294967295}
                propList={
                  Array [
                    "fill",
                  ]
                }
                r={16}
              />
              <RNSVGText
                alignmentBaseline="central"
                content={null}
                dx={Array []}
                dy={Array []}
                fill={4294967295}
                font={
                  Object {
                    "fontSize": "24",
                    "fontWeight": "bolder",
                    "textAnchor": "middle",
                  }
                }
                propList={
                  Array [
                    "fill",
                  ]
                }
                rotate={Array []}
                x={
                  Array [
                    "50%",
                  ]
                }
                y={
                  Array [
                    "50%",
                  ]
                }
              >
                <RNSVGTSpan
                  content="T"
                  dx={Array []}
                  dy={Array []}
                  font={Object {}}
                  rotate={Array []}
                  x={Array []}
                  y={Array []}
                />
              </RNSVGText>
            </RNSVGGroup>
          </RNSVGSvgView>
          <View
            style={
              Object {
                "flexBasis": "auto",
                "flexGrow": 1,
                "flexShrink": 1,
                "marginLeft": 12,
                "marginRight": 12,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "RegularFont",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "lineHeight": 24,
                  },
                  Array [
                    Object {
                      "fontFamily": "MediumFont",
                      "fontWeight": "500",
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 1)",
                    },
                  ],
                ]
              }
              testID="balances_row_2_symbol"
            />
            <Text
              ellipsizeMode="tail"
              numberOfLines={1}
              style={
                Array [
                  Object {
                    "fontFamily": "RegularFont",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "lineHeight": 24,
                  },
                  Array [
                    Object {
                      "color": "rgba(82, 82, 82, 1)",
                      "fontFamily": "MediumFont",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "lineHeight": 20,
                    },
                    Object {
                      "color": "rgba(82, 82, 82, 1)",
                    },
                  ],
                ]
              }
              testID="balances_row_2_name"
            >
              Ethereum
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "RegularFont",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "lineHeight": 24,
                  },
                  Array [
                    Object {
                      "flexWrap": "wrap",
                      "marginRight": 8,
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 1)",
                    },
                  ],
                ]
              }
              testID="balances_row_2_amount"
            >
              555.00000000
            </Text>
            <Text />
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
